silicon/backends/rabbitmq.hh

```diff
@@ -0,0 +1,260 @@
+	{
+		amqp_envelope_t envelope;
+	};
+
+	struct response
+	{
+		amqp_rpc_reply_t res;
+	};
+
+	struct service_utils
+	{
+		typedef request request_type;
+		typedef response response_type;
+
+		template <typename P, typename T>
+		auto
+		deserialize(request_type * r, P procedure, T & res) const
+		{
+
+			auto routing_key = get_string(r->envelope.routing_key);
+			auto message = get_string(r->envelope.message.body);
+			auto content_type = get_string(r->envelope.message.properties.content_type);
+			auto exchange = get_string(r->envelope.exchange);
+
+			//std::cout << "Delivery " << (unsigned) r->envelope.delivery_tag << " "
+			//		  << "exchange " << exchange << " "
+			//		  << "routingkey " << routing_key << " "
+			//		  << std::endl;
+
+			if (r->envelope.message.properties._flags & AMQP_BASIC_CONTENT_TYPE_FLAG)
+			{
+				//std::cout << "Content-type: " << content_type << std::endl;
+				//std::cout << "Message: " << message << std::endl;
+
+				iod::json_decode<typename P::route_type::parameters_type>(res, message);
+			}
+			//std::cout << "----" << std::endl;
+		}
+
+		template <typename T>
+		auto
+		serialize(response_type * r, T const & res) const
+		{
+		}
+	};
+
+	struct context
+	{
+		template <typename... O>
+		context(unsigned short port, O &&... opts)
+		{
+			auto options = D(opts...);
+			auto hostname = options.get(s::_hostname, std::string("localhost"));
+			auto username = options.get(s::_username, std::string("guest"));
+			auto password = options.get(s::_password, std::string("guest"));
+
+			conn = amqp_new_connection();
+			socket = amqp_tcp_socket_new(conn);
+
+			if (!socket)
+				throw std::runtime_error("creating TCP socket");
+
+			status = amqp_socket_open(socket, hostname.c_str(), port);
+			if (status)
+				throw std::runtime_error("opening TCP socket");
+
+			die_on_amqp_error(amqp_login(conn, "/", 0, 131072, 0, AMQP_SASL_METHOD_PLAIN, username.c_str(), password.c_str()), "Loggin in");
+			amqp_channel_open(conn, 1);
+			die_on_amqp_error(amqp_get_rpc_reply(conn), "Opening channel");
+		}
+
+		int status;
+		amqp_socket_t * socket = nullptr;
+		amqp_connection_state_t conn;
+		std::vector<amqp_bytes_t> queuenames;
+	};
+
+	template <typename A, typename M, typename... O>
+	auto
+	make_context(A const & api, M const & mf, unsigned short port, O &&... opts)
+	{
+		auto ctx = context(port, opts...);
+		auto options = D(opts...);
+		auto exchange = options.get(s::_exchange, std::string(""));
+
+		foreach(api) | [&] (auto& m)
+		{
+			iod::static_if<is_tuple<decltype(m.content)>::value>(
+					[&] (auto m) { // If sio, recursion.
+						throw std::runtime_error("FIXME: m.content is a tuple, not handle today");
+					},
+					[&] (auto m) { // Else, register the procedure.
+						std::stringstream bindingkey;
+
+						foreach(m.route.path) | [&] (auto e)
+						{
+
+							iod::static_if<is_symbol<decltype(e)>::value>(
+									[&] (auto e2) {
+										bindingkey << std::string("/") + e2.name();
```

Comment:
You should use m.route.string_id() to get the string representing the path of a route. https://github.com/matt-42/silicon/blob/master/silicon/http_route.hh#L220 This is what service use a a key to index the procedures.